src/reactivity/reactive.ts

```diff
@@ -203,29 +176,47 @@ export function shallowReactive<T extends object = any>(obj: T): T {
 
 export function markReactive(target: any, shallow = false) {
   if (
-    !isPlainObject(target) ||
+    !(isPlainObject(target) || Array.isArray(target)) ||
+    // !isPlainObject(target) ||
     isRaw(target) ||
-    Array.isArray(target) ||
+    // Array.isArray(target) ||
     isRef(target) ||
     isComponentInstance(target)
   ) {
     return
   }
 
-  if (
-    hasOwn(target, ReactiveIdentifierKey) &&
-    target[ReactiveIdentifierKey] === ReactiveIdentifier
-  ) {
+  if (isReactive(target) || !Object.isExtensible(target)) {
     return
   }
 
-  if (Object.isExtensible(target)) {
-    def(target, ReactiveIdentifierKey, ReactiveIdentifier)
-  }
+  reactiveSet.add(target)
 
   if (shallow) {
     return
   }
+
+  if (Array.isArray(target)) {
+    if (
+      // @ts-ignore
+      target.__ob__ &&
+      // @ts-ignore
+      target.__ob__.dep &&
+      // @ts-ignore
+      isFunction(target.__ob__.dep.addSub)
+    ) {
+      // @ts-ignore
+      target.__ob__.dep.addSub({
+        update() {
+          target.forEach((x) => markReactive(x))
+        },
+      })
+    }
+    // TODO way to track new array items
```

Comment:
Should we hijack the array methods or just leave a caveat in the readme?

Reply:
Is quite interesting idea, just thinking we might need to listen to other "events" to keep it consistent, for example, if you assign an array to an object, are we hijacking that array?

Reply:
What do you mean "events"? If I recall correctly, Vue 2 hijacked the array's push/unshift, etc to achieve the reactivity updates.

Reply:
when you add a new property to an object or when you change the property of the obj to an array

Reply:
Oh, I see, I got what you mean. And yeah, if we handle the hijacking here and wrap the value with `markReactive` for newly added properties, they should be well covered I think.